### PR TITLE
OCSP request parsing

### PR DESCRIPTION
**Pull request by reaperhulk**
Limited parsing support.

### CODE REVIEW

File: docs/development/test-vectors.rst [hunk 1]
```diff
@@ -409,6 +409,10 @@ Custom X.509 OCSP Test Vectors
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * ``x509/ocsp/req-sha1.der`` - An OCSP request containing a single request and
   using SHA1 as the hash algorithm.
+* ``x509/ocsp/req-multi-sha1.der`` - An OCSP request containing multiple
```
Comment (alex) [inline docs/development/test-vectors.rst]: Vectors to their own PR for ease of review please :-)

File: docs/x509/ocsp.rst [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    .. method:: public_bytes(encoding)
+
+        :param encoding: The encoding to use. Only
+            :attr:`~cryptography.hazmat.primitives.serialization.Encoding.DER`
+            is supported.
+
+        :return bytes: The serialized OCSP request.
+
+.. class:: Request
+
```
Comment (alex) [inline docs/x509/ocsp.rst]: Add a few words of prose describing it.

File: docs/x509/ocsp.rst [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            :class:`~cryptography.hazmat.primitives.hashes.Hash`
+
+        The algorithm used to generate the ``issuer_key_hash`` and
+        ``issuer_name_hash``.
+
+    .. attribute:: serial_number
+
+        :type: int
+
+        The serial number of the revoked certificate.
```
Comment (alex) [inline docs/x509/ocsp.rst]: The serial isn't necessarily revoked :-)

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 4]
```diff
@@ -1419,6 +1420,13 @@ def derive_elliptic_curve_private_key(self, private_value, curve):
 
         return _EllipticCurvePrivateKey(self, ec_cdata, evp_pkey)
 
+    def load_der_ocsp_request(self, data):
+        mem_bio = self._bytes_to_bio(data)
+        request = self._lib.d2i_OCSP_REQUEST_bio(mem_bio.bio, self._ffi.NULL)
+        self.openssl_assert(request != self._ffi.NULL)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: This is definitely capable of failing if you give it gibberish bytes, right?

File: src/cryptography/hazmat/backends/openssl/ocsp.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            self._backend._ffi.NULL, self._backend._ffi.NULL, self._cert_id
+        )
+        self._backend.openssl_assert(res == 1)
+        self._backend.openssl_assert(asn1obj[0] != self._backend._ffi.NULL)
+        oid = _obj2txt(self._backend, asn1obj[0])
+        try:
+            return _OIDS_TO_HASH[oid]
+        except KeyError:
+            raise UnsupportedAlgorithm(
+                "Signature algorithm OID:{0} not recognized".format(oid)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/ocsp.py]: Space after the colon

File: src/cryptography/x509/ocsp.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+from __future__ import absolute_import, division, print_function
+
+import abc
+
+import six
+
+from cryptography.hazmat.primitives import hashes
+
+
+_OIDS_TO_HASH = {
```
Comment (alex) [inline src/cryptography/x509/ocsp.py]: Why are these oids different from the x.509 ones... is it to make me mad?

File: src/cryptography/x509/ocsp.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+def load_der_ocsp_request(data):
+    from cryptography.hazmat.backends.openssl.backend import backend
+    return backend.load_der_ocsp_request(data)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class OCSPRequest(object):
+    @abc.abstractmethod
+    def __iter__(self):
+        """ Must support iteration of Requests """
```
Comment (alex) [inline src/cryptography/x509/ocsp.py]: Nit: we use multi-line docstrings everywhere else

File: tests/x509/test_ocsp.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    )
+
+
+class TestOCSPRequest(object):
+    def test_load_request_one_item(self):
+        req = _load_data(
+            os.path.join("x509", "ocsp", "req-sha1.der"),
+            ocsp.load_der_ocsp_request,
+        )
+        len(req) == 1
```
Comment (alex) [inline tests/x509/test_ocsp.py]: missing assert

File: tests/x509/test_ocsp.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        assert req[0].serial_number == int(
+            "98D9E5C0B4C373552DF77C5D0F1EB5128E4945F9", 16
+        )
+
+    def test_load_request_multiple_items(self):
+        req = _load_data(
+            os.path.join("x509", "ocsp", "req-multi-sha1.der"),
+            ocsp.load_der_ocsp_request,
+        )
+        len(req) == 2
```
Comment (alex) [inline tests/x509/test_ocsp.py]: assert

File: src/cryptography/x509/ocsp.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def load_der_ocsp_request(data):
+    from cryptography.hazmat.backends.openssl.backend import backend
+    return backend.load_der_ocsp_request(data)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class OCSPRequest(object):
+    @abc.abstractmethod
```
Comment (alex) [inline src/cryptography/x509/ocsp.py]: Should this have an `__len__`?

File: src/cryptography/hazmat/backends/openssl/ocsp.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+    def _request(self, idx):
+        request = self._backend._lib.OCSP_request_onereq_get0(
+            self._ocsp_request, idx
+        )
+        self._backend.openssl_assert(request != self._backend._ffi.NULL)
+        return _Request(self._backend, self._ocsp_request, request)
+
+    def public_bytes(self, encoding):
+        if encoding is not serialization.Encoding.DER:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/ocsp.py]: Is there a reason for this not to support PEM?

Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/ocsp.py]: Yes, OCSP requests are submitted either via DER (POST request) or via URL-encoded base64 encoding with no line breaks (GET). That form isn't really PEM. Maybe we should offer to serialize directly into that form, but I'd rather defer that to a subsequent PR.

**Pull request merged**